From 5116a49982cdc2f8abb6dbf1858560f3b1d1e098 Mon Sep 17 00:00:00 2001
From: Gabriel Fernandez <gabriel.fernandez@foss.st.com>
Date: Thu, 19 Oct 2023 14:00:50 +0200
Subject: [PATCH 0483/1695] clk: stm32mp25: set ck_icn_p_gicv2m as critical

The clock giv2m should be managed as system a clock and then
should be in critical. We have to restore it after a suspend.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git 197abd59f805ddf98ffb0d14e6f1e2be5fd80d34 clk: stm32mp25: set ck_icn_p_gicv2m as critical]
Signed-off-by: Gabriel Fernandez <gabriel.fernandez@foss.st.com>
Change-Id: Ibb5f44ecb87604c4235a099d591c8bbf9f9b788c
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/334889
Reviewed-by: Gatien CHEVALLIER <gatien.chevallier@st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Domain-Review: Patrick DELAUNAY <patrick.delaunay@foss.st.com>
---
 core/drivers/clk/clk-stm32-core.c | 15 +++++++++++++++
 core/drivers/clk/clk-stm32-core.h | 14 ++++++++++++++
 core/drivers/clk/clk-stm32mp25.c  |  2 +-
 3 files changed, 30 insertions(+), 1 deletion(-)

diff --git a/core/drivers/clk/clk-stm32-core.c b/core/drivers/clk/clk-stm32-core.c
index 7258ce7c0..98922aea9 100644
--- a/core/drivers/clk/clk-stm32-core.c
+++ b/core/drivers/clk/clk-stm32-core.c
@@ -375,6 +375,21 @@ const struct clk_ops clk_stm32_gate_ops = {
 	.is_enabled	= clk_stm32_gate_is_enabled,
 };
 
+static void clk_stm32_gate_pm_restore(struct clk *clk)
+{
+	struct clk_stm32_gate_cfg *cfg = clk->priv;
+
+	if (clk_is_enabled(clk))
+		stm32_gate_endisable(cfg->gate_id, true);
+}
+
+const struct clk_ops clk_stm32_gate_pm_ops = {
+	.enable		= clk_stm32_gate_enable,
+	.disable	= clk_stm32_gate_disable,
+	.is_enabled	= clk_stm32_gate_is_enabled,
+	.restore_context = clk_stm32_gate_pm_restore,
+};
+
 TEE_Result clk_stm32_gate_ready_enable(struct clk *clk)
 {
 	struct clk_stm32_gate_cfg *cfg = clk->priv;
diff --git a/core/drivers/clk/clk-stm32-core.h b/core/drivers/clk/clk-stm32-core.h
index 8eaff9797..1802f5b08 100644
--- a/core/drivers/clk/clk-stm32-core.h
+++ b/core/drivers/clk/clk-stm32-core.h
@@ -167,6 +167,8 @@ extern const struct clk_ops clk_stm32_divider_ops;
 extern const struct clk_ops clk_stm32_mux_ops;
 extern const struct clk_ops clk_stm32_composite_ops;
 
+extern const struct clk_ops clk_stm32_gate_pm_ops;
+
 #define PARENT(x...) { x }
 
 #define STM32_FIXED_RATE(_name, _rate)\
@@ -256,6 +258,18 @@ extern const struct clk_ops clk_stm32_composite_ops;
 		.parents = _parents,\
 	}
 
+#define STM32_GATE_PM(_name, _parent, _flags, _gate_id)\
+	struct clk _name = {\
+		.ops = &clk_stm32_gate_pm_ops,\
+		.priv = &(struct clk_stm32_gate_cfg) {\
+			.gate_id = _gate_id,\
+		},\
+		.name = #_name,\
+		.flags = (_flags),\
+		.num_parents = 1,\
+		.parents = { (_parent) },\
+	}
+
 struct clk_stm32_priv *clk_stm32_get_priv(void);
 uintptr_t clk_stm32_get_rcc_base(void);
 
diff --git a/core/drivers/clk/clk-stm32mp25.c b/core/drivers/clk/clk-stm32mp25.c
index 48576efc7..226b1cd67 100644
--- a/core/drivers/clk/clk-stm32mp25.c
+++ b/core/drivers/clk/clk-stm32mp25.c
@@ -3496,7 +3496,7 @@ static STM32_GATE(ck_icn_p_ltdc, &ck_icn_apb4, 0, GATE_LTDC);
 static STM32_GATE(ck_icn_p_csi2, &ck_icn_apb4, 0, GATE_CSI);
 static STM32_GATE(ck_icn_p_dcmipp, &ck_icn_apb4, 0, GATE_DCMIPP);
 static STM32_GATE(ck_icn_p_lvds, &ck_icn_apb4, 0, GATE_LVDS);
-static STM32_GATE(ck_icn_p_gicv2m, &ck_icn_apb4, 0, GATE_GICV2M);
+static STM32_GATE_PM(ck_icn_p_gicv2m, &ck_icn_apb4, 0, GATE_GICV2M);
 static STM32_GATE(ck_icn_p_usbtc, &ck_icn_apb4, 0, GATE_USBTC);
 static STM32_GATE(ck_icn_p_usb3pciephy, &ck_icn_apb4, 0, GATE_USB3PCIEPHY);
 static STM32_GATE(ck_icn_p_stgen, &ck_icn_apb4, 0, GATE_STGEN);
-- 
2.47.2

